fatrace.c

```diff
@@ -454,16 +469,71 @@ print_event (const struct fanotify_event_metadata *data,
                 problems[problem_idx++] = "path contains invalid UTF-8, path_raw contains the bytes.";
             }
         }
-        if (problem_idx) {
-          for (int i = 0; i < problem_idx; i++) {
-            printf("%s%s", i ? "\",\"" : ",\"problems\":[\"", problems[i]);
-          }
-          printf("\"]");
-        }
-        printf("}\n");
     } else {
-        printf ("%s(%i)%s: %-3s %s\n", procname[0] == '\0' ? "unknown" : procname, data->pid, printbuf, mask2str (data->mask), pathname);
+        printf ("%s(%i)%s: %-3s %s", procname[0] == '\0' ? "unknown" : procname, data->pid, printbuf, mask2str (data->mask), pathname);
+    }
+    if (option_ancestors && got_ppid) {
+        printf(option_json ? ",\"ancestors\":" : ", ancestors");
+        char sep = option_json ? '[' : '=';
+        bool problem_p_comm_not_found = false;
+        bool problem_p_comm_raw = false;
+        while (ppid) {
+            printf(option_json ? "%c{\"pid\":%i" : "%c(pid=%i", sep, ppid);
+            sep = ',';
+            snprintf (printbuf, sizeof (printbuf), "/proc/%i", ppid);
+            int ppid_dir_fd = open (printbuf, O_RDONLY | O_DIRECTORY);
+            if (ppid_dir_fd >= 0) {
+                char p_procname[TASK_COMM_LEN];
+                int p_procname_fd = openat (ppid_dir_fd, "comm", O_RDONLY);
+                ssize_t len = read (p_procname_fd, p_procname, sizeof (p_procname));
```

Comment:
Missing error check from open(). Please use a helper function for this.

Reply:
Was this addressed, or is it still a problem?

Reply:
Dang, I missed that during review of the other PR, sorry. `get_procname()` indeed is still missing that error check. It's *probably* harmless, `read()` ought to fail with `EBADF`, but you'll get a confusing/unhelpful error message.

Reply:
Opened #61.